### PR TITLE
[codex] fix responses lifecycle parity and codex session serialization

### PR DESCRIPTION
**Pull request by cita-777**
## Summary

This PR fixes the Responses lifecycle parity issues on the `codex/fix-responses-lifecycle-parity` branch and extends the fix to cover Codex HTTP session contention.

From the user side there were two classes of failures. One class was request-shape and lifecycle breakage when Responses traffic had to normalize or downgrade across `/v1/responses`, `/v1/messages`, and `/v1/chat/completions`: orphaned tool outputs, blank tool names, and Codex-specific request field mismatches could all surface as `400` errors. The second class was a Codex-specific concurrency failure where concurrent HTTP `/v1/responses` requests that reused the same `Session_id` could hit upstream session-slot contention and fail before any upstream path was recorded.

## Root Cause

The Responses compatibility pipeline had several places where lifecycle assumptions were too loose.

When Responses input or fallback payloads were converted between native Responses items and OpenAI-compatible message formats, the code did not consistently validate tool-call lifecycle pairs. That allowed blank function names to leak upstream, allowed orphaned `function_call_output` items to survive downgrade paths that require a matching assistant tool call, and left Codex-native request shaping incomplete for `store=false` handling.

Separately, Codex websocket transport already serialized requests per session, but plain HTTP `/v1/responses` traffic did not. Because Codex session continuity reuses `Session_id` across requests, overlapping HTTP turns for the same session could race each other upstream and hit the provider's session-slot limit.

## Fix

The patch makes the Responses path stricter where it must be strict and preserves continuity where the native API requires it.

Responses normalization now trims and validates tool lifecycle items, drops explicit blank tool names, and keeps follow-up tool outputs intact for native continuation cases. Compatibility conversion now filters all-invalid function tools, suppresses invalid tool choices when no valid tools remain, and removes orphaned tool outputs only when converting into OpenAI-compatible fallback formats that require a matching prior tool call.

Codex native Responses requests now always force `store=false`, matching the current upstream requirement. On top of that, HTTP Codex `/v1/responses` dispatch now runs through a lightweight in-process queue keyed by the final `Session_id`, so requests that share a Codex session are serialized locally the same way the websocket runtime already serializes websocket turns.

The proxy test transport was also updated to decode zstd-compressed buffered and non-SSE fallback responses through the shared runtime response reader so diagnostics and replay behavior stay consistent.

## Validation

I ran the following checks locally:

- `npm test -- src/server/transformers/openai/responses/conversion.test.ts src/server/routes/proxy/responses.codex-oauth.test.ts src/server/routes/proxy/chat.codex-oauth.test.ts src/server/routes/proxy/upstreamEndpoint.test.ts src/server/routes/proxy/chat.stream.test.ts src/server/routes/proxy/responses.websocket.test.ts src/server/routes/api/test.proxy.test.ts src/server/proxy-core/providers/registry.test.ts`

These cover the original Responses conversion regressions, Codex OAuth request shaping, fallback compatibility across chat/messages/responses, websocket continuation behavior, proxy test transport decoding, and the new regression that verifies concurrent HTTP Codex requests with the same session id are serialized before hitting upstream.

### CODE REVIEW

File: src/server/proxy-core/surfaces/openAiResponsesSurface.ts [hunk 1]
```diff
@@ -439,12 +445,11 @@ export async function handleOpenAiResponsesSurfaceRequest(
...
           };
+          const websocketTransportRequest = isResponsesWebsocketTransportRequest(request.headers as Record<string, unknown>);
           const streamSession = openAiResponsesTransformer.proxyStream.createSession({
             modelName,
             successfulUpstreamPath,
-            strictTerminalEvents: Object.entries(request.headers as Record<string, unknown>)
-              .some(([rawKey, rawValue]) => rawKey.trim().toLowerCase() === 'x-metapi-responses-websocket-transport'
-                && String(rawValue).trim() === '1'),
+            strictTerminalEvents: websocketTransportRequest,
             getUsage: () => parsedUsage,
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/proxy-core/surfaces/openAiResponsesSurface.ts:452]: Fixed in 05c7912. The websocket-transport fallback no longer forces native Responses terminal semantics onto downgraded chat-completions streams, and I added a regression test covering websocket transport replay of downgraded chat SSE on /v1/responses.

File: src/server/proxy-core/surfaces/openAiResponsesSurface.ts [hunk 2]
```diff
@@ -542,6 +547,61 @@ export async function handleOpenAiResponsesSurfaceRequest(
...
+                  upstreamPath: successfulUpstreamPath,
+                  runtimeFailureStatus: 502,
+                });
+                return;
+              }
+
+              await finalizeStreamSuccess(parsedUsage, latency);
+              return;
+            }
+          }
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/proxy-core/surfaces/openAiResponsesSurface.ts:605]: Fixed in 05c7912. After reading the SSE body up front for websocket transport, the fallback path now replays that same raw text through createSingleChunkStreamReader instead of attempting to re-read an already-consumed body stream.

File: src/server/routes/proxy/responsesSseFinal.ts [hunk 3]
```diff
@@ -120,6 +145,63 @@ function materializeCompletedPayloadFromAggregate(
...
+      next.output_text = derivedOutputText;
+    }
+  }
+
+  if (materialized && next.usage === undefined && materialized.usage !== undefined) {
+    next.usage = materialized.usage;
+  }
+
+  return next;
+}
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/routes/proxy/responsesSseFinal.ts:223]: Fixed in 05c7912. Terminal payload repair now preserves response.incomplete semantics instead of materializing completed-shaped aggregate payloads, and I added coverage for repairing empty incomplete payloads while keeping incomplete item status/details intact.

File: src/server/transformers/shared/chatFormatsCore.ts [hunk 4]
```diff
@@ -581,6 +611,78 @@ function collectToolCallsFromResponsesPayload(payload: Record<string, unknown>):
...
+
+  return {
+    toolCallDeltas: [{
+      index: outputIndex,
+      id: toolCallId,
+      name: toolName,
+      argumentsDelta: argumentsDelta || undefined,
+    }],
+  };
+}
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/transformers/shared/chatFormatsCore.ts:828]: Fixed in 05c7912. Responses tool-call streaming now uses a stable canonical tool-call index mapping keyed off output_index, item_id, and call_id so mixed output arrays and response.completed replays do not re-index or duplicate tool deltas.

File: src/server/routes/proxy/chat.stream.test.ts [hunk 5]
```diff
@@ -1154,6 +1154,56 @@ describe('chat proxy stream behavior', () => {
...
+    expect(response.headers['content-type']).toContain('text/event-stream');
+    expect(response.body).toContain('event: response.output_item.added');
+    expect(response.body).toContain('event: response.reasoning_summary_part.added');
+    expect(response.body).toContain('event: response.reasoning_summary_text.delta');
+    expect(response.body.indexOf('event: response.output_item.added')).toBeLessThan(
+      response.body.indexOf('event: response.reasoning_summary_part.added'),
+    );
+    expect(response.body.indexOf('event: response.reasoning_summary_part.added')).toBeLessThan(
+      response.body.indexOf('event: response.reasoning_summary_text.delta'),
+    );
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/routes/proxy/chat.stream.test.ts:1253]: Tightened in 05c7912. The test now specifically asserts that the parent event before reasoning summary events is a reasoning output item, not just any response.output_item.added frame.

File: src/server/routes/proxy/responses.websocket.test.ts [hunk 6]
```diff
@@ -572,20 +610,62 @@ describe('responses websocket transport', () => {
...
+      type: 'response.create',
+      model: 'gpt-5.4',
+      input: [],
+    }));
+
+    const message = await messagePromise;
+    socket.close();
+
+    expect(message?.response?.id).toBe('resp_http_incomplete');
   });
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/routes/proxy/responses.websocket.test.ts:676]: Tightened in 05c7912. The websocket test now keeps the socket open briefly after response.incomplete and explicitly asserts that no later error frame arrives before closing.

File: src/server/transformers/shared/chatFormatsCore.ts [hunk 7]
```diff
@@ -1335,20 +1588,105 @@ export function normalizeUpstreamStreamEvent(
...
     if (isNonEmptyString(responsePayload.model)) context.model = responsePayload.model;
+    const content = parseResponsesOutputText(responsePayload);
+    const contentDelta = computeNovelResponsesDelta(joinNonEmpty(Object.values(context.responsesTextByIndex)), content);
+    if (content) {
+      context.responsesTextByIndex = { ...context.responsesTextByIndex, [-1]: content } as Record<number, string>;
+    }
+    const responsesReasoning = parseResponsesReasoning(responsePayload);
+    const reasoningDelta = computeNovelResponsesDelta(
+      joinNonEmpty(Object.values(context.responsesReasoningByIndex)),
+      responsesReasoning.reasoningContent,
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/transformers/shared/chatFormatsCore.ts]: Fixed in ae84ebc. response.completed reconciliation now uses index-ordered exact segment joining instead of joinNonEmpty(Object.values(...)), so we preserve streamed whitespace and avoid injecting bogus terminal deltas. I also added a regression test covering streamed trailing whitespace before response.completed.

File: src/server/transformers/shared/chatFormatsCore.ts [hunk 8]
```diff
@@ -1238,19 +1473,33 @@ export function normalizeUpstreamStreamEvent(
...
-      ? payload.delta
-      : extractTextAndReasoning(payload.delta).content;
+  if (type === 'response.reasoning_summary_text.delta' || type === 'response.reasoning_summary_text.done') {
+    const outputIndex = extractResponsesOutputIndex(payload);
+    const deltaText = type === 'response.reasoning_summary_text.done'
+      ? (typeof (payload as any).text === 'string' ? (payload as any).text : extractTextAndReasoning(payload.text).content)
+      : (typeof payload.delta === 'string' ? payload.delta : extractTextAndReasoning(payload.delta).content);
+    const novelDelta = computeNovelResponsesDelta(context.responsesReasoningByIndex[outputIndex] || '', deltaText);
+    context.responsesReasoningByIndex[outputIndex] = deltaText || context.responsesReasoningByIndex[outputIndex] || '';
     return {
```
[the remark this answers is not in the document]
Comment (cita-777) [reply src/server/transformers/shared/chatFormatsCore.ts:1525]: Fixed in ae84ebc. reasoning summary stream state now accumulates prior delta chunks and treats *.done as the canonical final text, so later done events no longer replay the whole reasoning summary after multiple delta frames. Added a regression test for delta + delta + done.
